src/kernel/filesystem/fat32.zig

```diff
@@ -0,0 +1,1756 @@
+        ///
+        /// Return: *vfs.Node
+        ///     The VFS Node for the opened file/folder.
+        ///
+        /// Error: Allocator.Error || ReadError || SeekError || vfs.Error
+        ///     Allocator.Error           - Not enough memory for allocating memory
+        ///     ReadError                 - Errors when reading the stream.
+        ///     SeekError                 - Errors when seeking the stream.
+        ///     vfs.Error.NoSuchFileOrDir - Error if the file/folder doesn't exist.
+        ///
+        fn openImpl(fs: *const vfs.FileSystem, dir: *const vfs.DirNode, name: []const u8) (Allocator.Error || ReadError || SeekError || vfs.Error)!*vfs.Node {
+            const self = @fieldParentPtr(Fat32Self, "instance", fs.instance);
+
+            // TODO: Future PR
+            return vfs.Error.NoSuchFileOrDir;
+        }
+
+        ///
+        /// Helper function for creating a file, folder or symlink.
+        ///
+        /// Arguments:
+        ///     IN fs: *const vfs.FileSystem - The underlying filesystem.
+        ///     IN dir: *const vfs.DirNode   - The parent directory.
+        ///     IN name: []const u8          - The name of the file/folder to open.
+        ///     IN flags: vfs.OpenFlags      - The open flags for if creating a file/folder/symlink.
+        ///     IN open_args: vfs.OpenArgs   - The open arguments if creating a symlink.
+        ///
+        /// Return: *vfs.Node
+        ///     The VFS Node for the opened/created file/folder.
+        ///
+        /// Error: Allocator.Error || ReadError || SeekError || vfs.Error
+        ///     Allocator.Error           - Not enough memory for allocating memory
+        ///     ReadError                 - Errors when reading the stream.
+        ///     SeekError                 - Errors when seeking the stream.
+        ///     vfs.Error.NoSuchFileOrDir - Error if creating a symlink and no target is provided.
+        ///
+        fn createFileOrDirOrSymlink(fs: *const vfs.FileSystem, dir: *const vfs.DirNode, name: []const u8, flags: vfs.OpenFlags, open_args: vfs.OpenArgs) (Allocator.Error || ReadError || SeekError || vfs.Error)!*vfs.Node {
+            const self = @fieldParentPtr(Fat32Self, "instance", fs.instance);
+
+            // TODO: Future PR
+            return vfs.Error.NoSuchFileOrDir;
+        }
+
+        ///
+        /// Deinitialise this file system.
+        ///
+        /// Arguments:
+        ///     IN self: *Fat32Self - Self to free.
+        ///
+        pub fn deinit(self: *Fat32Self) void {
+            // Make sure we have closed all files
+            // TODO: Should this deinit close any open files instead?
+            std.debug.assert(self.opened_files.count() == 0);
+            self.opened_files.deinit();
+            self.allocator.destroy(self.root_node.node);
+            self.allocator.destroy(self.fs);
+            self.allocator.destroy(self);
```

Comment:
In general I don't think deinit functions should destroy the struct itself since that would mean you couldn't deinit stack-allocated structs.